go/fn/run.go

```diff
@@ -28,40 +28,64 @@ import (
 // function and write the updated resourceList in yaml to stdout. Errors if any
 // will be printed to stderr.
 func AsMain(p ResourceListProcessor) error {
-	in, err := ioutil.ReadAll(os.Stdin)
-	if err != nil {
-		return fmt.Errorf("unable to read from stdin: %v", err)
-	}
-	out, err := Run(p, in)
-	if err != nil {
+	err := func() error {
+		in, err := ioutil.ReadAll(os.Stdin)
+		if err != nil {
+			return fmt.Errorf("unable to read from stdin: %v", err)
+		}
+		out, err := Run(p, in)
+		// If there is an error, we don't return the error immediately.
+		// We write out to stdout before returning any error.
+		_, outErr := os.Stdout.Write(out)
+		if outErr != nil {
+			return outErr
+		}
 		return err
-	}
-	_, err = os.Stdout.Write(out)
+	}()
 	return err
 }
 
 // Run evaluates the function. input must be a resourceList in yaml format. An
 // updated resourceList will be returned.
-func Run(p ResourceListProcessor, input []byte) ([]byte, error) {
+func Run(p ResourceListProcessor, input []byte) (out []byte, err error) {
 	rl, err := ParseResourceList(input)
 	if err != nil {
 		return nil, err
 	}
-	defer handleOptOrDieErr()
-	err = p.Process(rl)
-	if err != nil {
-		// If the error is not a Results type, we wrap the error as a Result.
-		if results, ok := err.(Results); ok {
-			rl.Results = append(rl.Results, results...)
-		} else if result, ok := err.(Result); ok {
-			rl.Results = append(rl.Results, &result)
-		} else if result, ok := err.(*Result); ok {
-			rl.Results = append(rl.Results, result)
-		} else {
-			rl.Results = append(rl.Results, ErrorResult(err))
+	defer func() {
+		// if we run into a panic, we still need to log the error to Results,
+		// and return the ResourceList and error.
+		v := recover()
+		if v != nil {
+			if e, ok := v.(ErrOpOrDie); ok {
+				err = fmt.Errorf(e.Error())
+				rl.LogResult(err)
+				out, _ = rl.ToYAML()
+			} else if e, ok := v.(Result); ok {
+				err = fmt.Errorf(e.Message)
+				rl.LogResult(err)
```

Comment:
If we convert a `Result` or `Results` error to an untyped error. We will lose some structured error information when the untyped error is passed to `LogResult`.

Reply:
That makes sense, thanks. I've changed it to your above suggestion.